### PR TITLE
[🚧 WIP 🚧] 🎉 DPP Threads Support

### PR DESCRIPTION
**Pull request by webtax-gh**
Threads are now partially added to DPP. They are basically channels, but with these new keys: 
image

Todo list:
[ ] Add thread specific gw events (currently threads are only imported through `guild_create`)
[ ] Add thread creation utils
[ ] Add thread REST utils

### CODE REVIEW

File: src/dpp/events/guild_create.cpp [hunk 1]
```diff
@@ -83,6 +83,20 @@ void guild_create::handle(discord_client* client, json &j, const std::string &ra
 			g->channels.push_back(c->id);
 		}
 
+		/* Store guild threads */
+		g->threads.clear();
+		g->threads.reserve(d["threads"].size());
+		for (auto & channel : d["threads"]) {
+			dpp::channel* c = dpp::find_channel(SnowflakeNotNull(&channel, "id"));
```
Comment (braindigitalis) [inline src/dpp/events/guild_create.cpp:90]: caching here needs to be optional, note without the caching we dont need the `new` at all

File: include/dpp/channel.h [hunk 2]
```diff
@@ -123,6 +147,15 @@ class channel : public managed {
 	/** Permission overwrites to apply to base permissions */
 	std::vector<permission_overwrite> permission_overwrites;
 
+	/** Approximate count of messages in a thread (threads) */
+	uint8_t message_count;
```
Comment (braindigitalis) [inline include/dpp/channel.h:151]: is this really big enough? perhaps uint32_t?

File: include/dpp/channel.h [hunk 3]
```diff
@@ -123,6 +147,15 @@ class channel : public managed {
...
+
+	/** Approximate count of members in a thread (threads) */
+	uint8_t member_count;
```
Comment (braindigitalis) [inline include/dpp/channel.h:154]: Is this big enough? does discord allow > 255 members in a thread? perhaps uint16_t?

File: include/dpp/channel.h [hunk 4]
```diff
@@ -123,6 +147,15 @@ class channel : public managed {
 	/** Permission overwrites to apply to base permissions */
 	std::vector<permission_overwrite> permission_overwrites;
 
+	/** Approximate count of messages in a thread (threads) */
+	uint8_t message_count;
+
+	/** Approximate count of members in a thread (threads) */
```
Comment (braindigitalis) [inline include/dpp/channel.h:153]: Remember to init both of these to 0 in channel::channel

**Review by braindigitalis — CHANGES_REQUESTED** [review]

File: include/dpp/channel.h (hunk 2 above)
Comment (webtax-gh) [reply include/dpp/channel.h:151]: yes, it's big enough 
image

File: include/dpp/channel.h (hunk 3 above)
Comment (webtax-gh) [reply include/dpp/channel.h:154]: yes, it's big enough 
image

File: include/dpp/channel.h (hunk 4 above)
Comment (webtax-gh) [reply include/dpp/channel.h:153]: resolved

**Comment by braindigitalis** [comment]
hi how far did this get, I was holding off on the next release until we had threads support

**Comment by braindigitalis** [comment]
continued and merged as part of PR#46

**Pull request closed without merging**